src/irmin-pack/inode.ml

```diff
@@ -527,17 +546,19 @@ struct
       if off < 0 then invalid_arg "Invalid pagination offset";
       if len < 0 then invalid_arg "Invalid pagination length";
       if len = 0 then Seq.empty
-      else fun () -> seq_v layout t.v ~cache empty_continuation ~off ~len
+      else fun () ->
+        seq_v layout t.v ~depth:0 ~cache empty_continuation ~off ~len
 
     let seq_tree layout ?(cache = true) i : (step * value) Seq.t =
       let off = 0 in
       let len = Int.max_int in
-      fun () -> seq_v layout (Tree i) ~cache empty_continuation ~off ~len
+      fun () ->
+        seq_v layout (Tree i) ~depth:0 ~cache empty_continuation ~off ~len
 
     let seq_v layout ?(cache = true) v : (step * value) Seq.t =
       let off = 0 in
       let len = Int.max_int in
-      fun () -> seq_v layout v ~cache empty_continuation ~off ~len
+      fun () -> seq_v layout v ~depth:0 ~cache empty_continuation ~off ~len
```

Comment:
There are now 11 occurences of `~depth:0` in the file, I think that only 3 of these are really safe (the ones protected by `check_write_op_supported`).

In the past we've made the assumption that the functions in `Val_impl` can be called on any inode and that the functions in `Val` can only be called on root inodes. It was an important distinction for the layered store, because of the `pred` function.

It makes these `~depth:0` a bit unsafe.

I think that these `~depth:0` should all be moved to `Val`, I'll check into more details this afternoon.

Reply:
I have double checked and it seems that we instantiate `Val.t`s that are non root during the freeze of the layered store: https://github.com/mirage/irmin/blob/main/src/irmin-pack/_layered/ext_layered.ml#L514-L520, https://github.com/mirage/irmin/blob/main/src/irmin-pack/inode.ml#L419-L421

In `main` (where the layered store is disabled) we could make it illegal to instantiate `Val.t` that aren't roots, this would make my above solution possible.

Another solution would be to fetch the right starting `depth` from the structure of the inode, instead of always picking `0`.